linter.py

```diff
@@ -1,57 +1,49 @@
 #
 # linter.py
-# Linter for SublimeLinter3, a code checking framework for Sublime Text 3
+# Linter for SublimeLinter4, a code checking framework for Sublime Text 3
 #
 # Written by Noah Davis
 # Copyright (c) 2016 Noah Davis
 #
+# Updated by Schuyler Jager & Andreas for SublimeLinter 4
+# Copyright (c) 2020 Schuyler Jager & Andreas
 # License: MIT
 #
 
 """This module exports the Codeclimate plugin class."""
-
-from SublimeLinter.lint import Linter, util, persist
+from SublimeLinter.lint import Linter, util
 
 
 class Codeclimate(Linter):
     """Provides an interface to codeclimate."""
-
-    syntax = (
-        'ruby on rails',
-        'ruby',
-        'javascript',
-        'python',
-        'go',
-        'html',
-        'css',
-        'markdown',
-        'php',
-        'haskall',
-        'java',
-        'shellscript'
-    )
-    executable = "codeclimate"
+    defaults = {
+        'selector': (
+            'source.css, '
+            'source.go, '
+            'source.haskell, '
+            'source.java, '
+            'source.javascript, '
+            'source.js, '
+            'source.php, '
+            'source.python, '
+            'source.ruby, '
+            'source.scss, '
+            'source.shell, '
+            'text.html'
+        )
+    }
     regex = r'^(?P<line>\d+)(?:-\d+)?:\s(?P<message>.+)$'
     multiline = False
     line_col_base = (1, 1)
     tempfile_suffix = '-'
     error_stream = util.STREAM_BOTH
-    selectors = {}
     word_re = None
-    defaults = {
-        'chdir': "${project}"
-    }
-    inline_settings = None
-    inline_overrides = None
-    comment_re = None
 
     def cmd(self):
-        """Construct a cmd to provide a relative path to 'codeclimate analyze'."""
-        result = [self.executable_path, 'analyze']
-        relative_file_name = None
-        for folder in self.view.window().folders():
-            if self.filename.find(folder) == 0:
-                relative_file_name = self.filename.replace(folder + '/', '')
-        result.append(relative_file_name)
-        persist.debug(result)
-        return result
+        """Set command and run 'codeclimate analyze'."""
+        if 'executable' in self.settings:
+            command = self.settings.get('executable')
+        else:
+            command = 'codeclimate'
+
+        return [command, 'analyze', '${args}', '${file}']
```

Comment:
Now it's so easy you don't have to define a method at all.  `executable` is handled by the SL framework.  

Just define `cmd = ('codeclimate', 'analyze', '${args}', '${file_on_disk}')`, maybe right before `defaults` for readability.  `file_on_disk` holds the same value as `file` but also acts as a marker for SublimeLinter.

Reply:
Hi @kaste

Thank you very much for your feedback.

Yes, that looks very easy. However, I noticed an additional issue. In some cases, I had results where shown in the wrong file – something like "the linter is showing File A's results in File B."
In short, I believe this effect appears because the file path is absolute and not relative to the working directory. (If you want to drive deeper into this, I believe https://github.com/codeclimate/codeclimate/issues/845 gives an overview of some of the key points.)

However, I solved the issue in doing this (short version without exception handling):

```python
path = os.path.relpath(self.filename, self.get_working_dir())
return ['codeclimate', 'analyze', '${args}', path]
```

But SublimeLinter shows me additionally the absolute path of the file at the end:

```text
INFO:SublimeLinter.lint.linter:Running ...

  /Users/run/Desktop/TEST  (working dir)
  $ /usr/local/bin/codeclimate analyze src/Plugin/Field/FieldWidget/ACLConfigSelector.php /Users/run/Desktop/TEST/src/Plugin/Field/FieldWidget/ACLConfigSelector.php
```

Even it is working, it makes me nervous in a way. Is this only a logging message?

I tried several steps to find a way to change this, but I didn't find a working solution. (I also searched for a function like `set_relative_path` or similar in the SublimeLinter code.) What would you do?

Reply:
Ah, the automatic filename appending... That's a shitty part of SublimeLinter we couldn't get rid of for compatibility reasons.  We probably can change that in 2021. 

We need a marker in the `cmd`, iirc `return ["codeclimate", ..., "${xoo}"]` should work.  (T.i. *any* unused, undefined marker.)

Reply:
Hi @kaste

Thank you very much for your answer. I'm not sure if I understand you correctly; or if my Python skills are not strong enough, to handle this "marker" thing.

I tried

```python
    def cmd(self):
        file = os.path.relpath(self.filename, self.get_working_dir())

        return ['codeclimate', 'analyze', '${args}', file, '${xoo}']
```

and, surprisingly, this works. But it looks to me horribly wrong. May you have a working example for me?

Reply:
Yeah, it's kinda horrible but easy to use.  When we see a marker like `$file_on_disk` or anything else we do *not* append the filename automatically.  When you use `$xoo` it will resolve to the empty string so this hack works afair.  (Note that this plugin is the only plugin using this hack, all other plugins either use `file_path` or `file_on_disk` as all of them work with absolute file paths.)

The auto-appending code is actually in the `finalize_cmd` method but there is no need to implement/override this method as it's more complicated and brittle I think.

Reply:
Hi @kaste

Oh, ok… I'm relieved.

I updated the command and appended the marker `'${xoo}'`. I also tried to catch someone trying to run the linter against a file from outside the working directory*. For this purpose, I check with [`if not (abs_path.startswith(os.path.abspath(working_dir) + os.sep))`][1]  if the current file is in the working directory. If it is not, I try to print a `logger.info` message and preventing SublimeLinter from linting the current file (`return None`). As far as I can see, it works but raises a backend error. Can I solve this case more accurately? 

I plan to eliminate the limitation in further releases, and I see several solutions and a bunch of work. For now, I only try to get this plugin to work together with SublimeLinter 4. 

[1]:https://github.com/codeclimate/SublimeLinter-contrib-codeclimate/pull/3/files#diff-e7c2c765d5c484aae3f449e93444a2659ddc4bac3b83d10c6ee5c36f6c9e8eebR67

_* I'm doing this because of the "the linter is showing File A's results in File B"-issue when codeclimate is called with an absolute file path or from outside the working directory._